ruby/ext/google/protobuf_c/map.c

```diff
@@ -0,0 +1,894 @@
+// their own bytes) and integers as sequences of either 4 or 8 bytes in
+// host-byte-order as either a uint32_t or a uint64_t.
+
+// Forms a key to use with the underlying strtable from a Ruby key value. |buf|
+// must point to TABLE_KEY_BUF_LENGTH bytes of temporary space, used to
+// construct a key byte sequence if needed. |out_key| and |out_length| provide
+// the resulting key data/length.
+#define TABLE_KEY_BUF_LENGTH 8  // sizeof(uint64_t)
+static void table_key(Map* self, VALUE key,
+                      char* buf,
+                      const char** out_key,
+                      size_t* out_length) {
+  switch (self->key_type) {
+    case UPB_TYPE_BYTES:
+    case UPB_TYPE_STRING:
+      // Strings: use string content directly.
+      Check_Type(key, T_STRING);
+      native_slot_validate_string_encoding(self->key_type, key);
+      *out_key = RSTRING_PTR(key);
+      *out_length = RSTRING_LEN(key);
+      break;
+
+    case UPB_TYPE_BOOL:
+    case UPB_TYPE_INT32:
+    case UPB_TYPE_INT64: {
+      // Signed numeric types: use an int64 in host-native byte order.
+      int64_t key_val = 0;
+
+      // Do a range/value check.
+      switch (self->key_type) {
+        case UPB_TYPE_BOOL:
+          if (key != Qtrue && key != Qfalse) {
+            rb_raise(rb_eTypeError, "Key must be true or false");
+          }
+          key_val = (key == Qtrue) ? 1 : 0;
+          break;
+        case UPB_TYPE_INT32:
+          native_slot_check_int_range_precision(self->key_type, key);
+          key_val = NUM2INT(key);
+          break;
+        case UPB_TYPE_INT64:
+          native_slot_check_int_range_precision(self->key_type, key);
+          key_val = NUM2LL(key);
+          break;
+        default:
+          break;
+      }
+
+      int64_t* int64_key = (int64_t*)buf;
+      *int64_key = key_val;
+      *out_key = buf;
+      *out_length = sizeof(int64_t);
+      break;
+    }
+
+    case UPB_TYPE_UINT32:
+    case UPB_TYPE_UINT64: {
+      // Unsigned numeric types: use a uint64 in host-native byte order.
+      uint64_t key_val = 0;
+
+      // Do a range/value check.
+      native_slot_check_int_range_precision(self->key_type, key);
+      switch (self->key_type) {
+        case UPB_TYPE_UINT32:
+          key_val = NUM2UINT(key);
+          break;
+        case UPB_TYPE_UINT64:
+          key_val = NUM2ULL(key);
+          break;
+        default:
+          break;
+      }
+
+      uint64_t* uint64_key = (uint64_t*)buf;
+      *uint64_key = key_val;
+      *out_key = buf;
+      *out_length = sizeof(uint64_t);
+      break;
+    }
+
+    default:
+      // Map constructor should not allow a Map with another key type to be
+      // constructed.
+      assert(false);
+      break;
+  }
+}
+
+static VALUE table_key_to_ruby(Map* self, const char* buf, size_t length) {
+  switch (self->key_type) {
+    case UPB_TYPE_BYTES:
+    case UPB_TYPE_STRING: {
+      VALUE ret = rb_str_new(buf, length);
```

Comment:
It's probably worth noting some tradeoffs of this approach:

+ parsing doesn't have to create a Ruby string object
- traversing a Map with string keys creates new strings every time: more copying, more garbage.

We may want to consider at some point whether string keys should store (or at least cache lazily-built) corresponding Ruby strings.